man/memkind_default.3
.\"
.\" Copyright (C) 2014 - 2019 Intel Corporation.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" 1. Redistributions of source code must retain the above copyright notice(s),
.\"    this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright notice(s),
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
.\" EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH "MEMKIND_DEFAULT" 3 "2015-04-21" "Intel Corporation" "MEMKIND_DEFAULT" \" -*- nroff -*-
.SH "NAME"
memkind_default \- default implementations for memkind operations.
.br
.BR Note:
This is EXPERIMENTAL API. The functionality and the header file itself can be changed (including non-backward compatible changes), or removed.
.SH "SYNOPSIS"
.nf
.B #include <memkind/internal/memkind_default.h>
.sp
.B Link with -lmemkind
.sp
.BI "int memkind_default_create(struct memkind " "*kind" ", struct memkind_ops " "*ops" ", const char " "*name" );
.br
.BI "int memkind_default_destroy(struct memkind " "*kind" );
.br
.BI "void *memkind_default_malloc(struct memkind " "*kind" ", size_t " "size" );
.br
.BI "void *memkind_default_calloc(struct memkind " "*kind" ", size_t " "num" ", size_t " "size" );
.br
.BI "int memkind_default_posix_memalign(struct memkind " "*kind" ", void " "**memptr" ", size_t " "alignment" ", size_t " "size" );
.br
.BI "void *memkind_default_realloc(struct memkind " "*kind" ", void " "*ptr" ", size_t " "size" );
.br
.BI "void memkind_default_free(struct memkind " "*kind" ", void " "*ptr" );
.br
.BI "void *memkind_default_mmap(struct memkind " "*kind" ", void " "*addr" ", size_t " "size" );
.br
.BI "int memkind_default_mbind(struct memkind " "*kind" ", void " "*ptr" ", size_t " "len" );
.br
.BI "int memkind_default_get_mmap_flags(struct memkind " "*kind" ", int " "*flags" );
.br
.BI "int memkind_default_get_mbind_mode(struct memkind " "*kind" ", int " "*mode" );
.br
.BI "size_t memkind_default_malloc_usable_size(struct memkind " "*kind" ", void " "*ptr" );
.br
.BI "int memkind_preferred_get_mbind_mode(struct memkind " "*kind" ", int " "*mode" );
.br
.BI "int memkind_interleave_get_mbind_mode(struct memkind " "*kind" ", int " "*mode" );
.br
.BI "int memkind_nohugepage_madvise(struct memkind " "*kind" ", void " "*addr" ", size_t " "size" );
.br
.BI "int memkind_posix_check_alignment(struct memkind " "*kind" ", size_t " "alignment" );
.br
.BI "int memkind_default_get_mbind_nodemask(struct memkind " "*kind" ", unsigned long " "*nodemask" ", unsigned long " "maxnode" );
.br
.BI "void memkind_default_init_once(void);"
.br
.BI "int memkind_default_get_defrag_hint(void " "*ptr" ", int " "*bin_util" ", int " "*run_util" );
.br
.BI "bool size_out_of_bounds(size_t " "size" );
.br
.SH DESCRIPTION
.PP
Default implementations for memkind operations which include a several
useful methods that are not part of the
.B MEMKIND_DEFAULT
kind which is a fall through to the jemalloc implementation.
.PP
.BR memkind_default_create ()
implements the required start up for every kind. If a kind does not
point to this function directly for its
.BR ops.create ()
operation, then the function that it points to must call
.BR memkind_default_create ()
at its start.
.PP
.BR memkind_default_destroy ()
implements the required shutdown for every kind. If a kind does not
point to this function directly for its
.BR ops.destroy ()
operation, then the function that it points to must call
.BR memkind_default_destroy ()
at its end.
.PP
.BR memkind_default_malloc ()
is a direct call through the jemalloc's
.BR malloc ().
.PP
.BR memkind_default_calloc ()
is a direct call through the jemalloc's
.BR calloc ().
.PP
.BR memkind_default_posix_memalign ()
is a direct call through the jemalloc's
.BR posix_memalign ().
.PP
.BR memkind_default_realloc ()
is a direct call through the jemalloc's
.BR realloc ().
.PP
.BR memkind_default_free ()
is a direct call through the jemalloc's
.BR free ().
Note that this method can be called on any pointer returned by a
jemalloc allocation, and in particular, all of the arena
allocations described in
.BR memkind_arena (3)
can use this function for freeing.
.PP
.BR memkind_default_mmap ()
This calls the ops->get_mmap_flags()
operations for the kind, or falls back on the default implementations
if the function pointers are
.IR "NULL" .
The results of these calls are
passed to the
.BR mmap (2)
call to allocate pages from the operating system.  The
.I addr
is the hint passed through to
.BR mmap (2)
and
.I size
is the size of the buffer to be allocated.  The return value is the
allocated buffer or
.B MAP_FAILED
in the case of an error.
.PP
.BR memkind_default_mbind ()
makes calls the kind's
.BR ops.get_mbind_nodemask ()
and
.BR ops.get_mbind_mode ()
operations to gather inputs and then calls the
.BR mbind (2)
system call using the results along with and user input
.I ptr
and
.IR len .
.PP
.BR memkind_default_get_mmap_flags ()
sets
.I flags
to
.BR "MAP_PRIVATE | MAP_ANONYMOUS" .
See
.BR mmap (2)
for more information about these flags.
.PP
.BR memkind_default_get_mbind_mode ()
sets
.I mode
to
.BR MPOL_BIND .
See
.BR mbind (2)
for more information about this flag.
.PP
.BR memkind_default_malloc_usable_size ()
is a direct call through the jemalloc's
.BR malloc_usable_size ().
.PP
.BR memkind_preferred_get_mbind_mode ()
sets
.I mode
to
.BR MPOL_PREFERRED .
See
.BR mbind (2)
for more information about this flag.
.PP
.BR memkind_interleave_get_mbind_mode ()
sets
.I mode
to
.BR MPOL_INTERLEAVE .
See
.BR mbind (2)
for more information about this flag.
.PP
.BR memkind_nohugepage_madvise ()
calls
.BR madvise (2)
with the
.B MADV_NOHUGEPAGE
advice.
See
.BR madvise (2)
for more information about this option.
.PP
.BR memkind_posix_check_alignment ()
can be used to check the alignment value for
.BR memkind_posix_memalign ()
to ensure that is abides by the POSIX requirements:
alignment must be a power of 2 at least as large as
.IR "sizeof(void*)" .
.PP
.BR memkind_default_get_mbind_nodemask ()
wraps jemalloc's
.BR copy_bitmask_to_bitmask .
This function copies body of the bitmask structure into passed pointer.
.PP
.BR memkind_default_init_once ()
initializes heap manager.
.PP
.BR memkind_default_get_defrag_hint ()
is a direct call through the jemalloc's
.BR get_defrag_hint ().
.PP
.BR size_out_of_bounds ()
returns true if given size is out of bounds, otherwise will return false.
.SH "COPYRIGHT"
Copyright (C) 2014 - 2019 Intel Corporation. All rights reserved.
.SH "SEE ALSO"
.BR memkind (3),
.BR memkind_arena (3),
.BR memkind_hbw (3),
.BR memkind_hugetlb (3),
.BR memkind_pmem (3),
.BR jemalloc (3),
.BR mbind (2),
.BR mmap (2)


ChangeLog
* Mon Dec 10 2018 Michal Biesek <michal.biesek@intel.com> v1.8.0
- Fixed error with realloc/free method with passing thread-specific cache flag
- Fixed error with memkind_create_pmem(), if other PMEM kind was destroyed before
- Fixed error with zeroing large allocations in PMEM kind
- Added support to create kind without maximum size limit of PMEM kind (max_size=0)
- Extend memkind API with memkind_malloc_usable_size()
- Removed EXPERIMENTAL from most methods in memkind API
- Added MEMKIND_ERROR_ARENAS_CREATE code
- Added C++ allocator for PMEM kind
- Extend PMEM examples
- Fixed integration with Travis CI
- Extend Travis CI configuration with Astyle and Coverage
- Added PMEM kind tests

* Wed Jan 24 2018 Pawel Kochanek <pawel.b.kochanek@intel.com> v1.7.0
- Updated internally used jemalloc to upstream version 5.0.
- Fixed error that has been occuring while memkind was dynamically loaded.
- Fixed MEMKIND_HBW_NODES behavior for single NUMA node system.
- Removed licenses other than BSD 3-clause from COPYING.
- Changed build instructions.
- Added configurable jemalloc prefix in build scripts.
- Upgraded gtest to version 1.8.0.
- Added memory footprint tests.
- Added locality test for MEMKIND_PREFERRED.
- Applied test parametrization in BATests.
- Fixed problems with pytest testing in Travis.
- Added huge page configuration in several tests.
- Removed several symbols that are no longer exposed in public API from man pages.
- Fixed HBW_POLICY_BIND_ALL documentation.

* Tue Jun 13 2017 Artur Koziej <artur.koziej@intel.com> v1.6.0
- Introduced new policy HBW_POLICY_BIND_ALL.
- Introduced MEMKIND_HBW_ALL, MEMKIND_HBW_ALL_HUGETLB and MEMKIND_REGULAR kinds.
- Fixed hbw_posix_memalign_psize() return value in case of unsupported policy and
  page size combination.
- Documented supported policy and page size combination for
  hbw_posix_memalign_psize().
- Merged libmemkind.a and libjemalloc_pic.a into a single static library.
- Fixed OOM during tcache acquisition.
- Tuned huge pages configuration in memkind tests.
- Added stress tests for allocations with large random sizes.
- Added stress tests for allocations with random kinds.
- Extended policy and NUMA node binding tests.
- Fixed several Travis build problems.
- Removed gb_page_tests_preferred_policy tests.

* Tue Feb 21 2017 Krzysztof Kulakowski <krzysztof.kulakowski@intel.com> v1.5.0
- Intel(R) TBB scalable_allocator can be used as heap management alternative
  to jemalloc (requires installed Intel (R) TBB package and environment variable
  MEMKIND_HEAP_MANAGER=TBB)
- Updated internally used jemalloc to upstream version 4.3.1 (previously modified 3.5.1)
- Introduced new environment option MEMKIND_HOG_MEMORY. Setting it to "1" will prevent
  memkind from releasing memory to OS.
- Introduced support for KNM processor from Intel(R) Xeon Phi(TM) family
- General cleanup of build scripts (e.g. fixed issues found on Ubuntu,
  introduced simple build.sh script)
- Deprecated support for gigabyte-pages
- Removed number of previously deprecated symbols

* Thu Nov 17 2016 Krzysztof Kulakowski <krzysztof.kulakowski@intel.com> v1.4.0
- Introduced hbw_verify_memory_region() for verifying if provided memory
  region fully falls into high-bandwidth memory (for details please take a look
  at man/hbwmalloc.3).
- New API for creating kinds using set of predefined attributes:
  memkind_create_kind(), memkind_destroy_kind() (for details please take a look
  at man/memkind.3).
- Fixed issue in autohbw where it was reporting limit error regardless of
  value provided.
- Added logging on error paths.
- Test base fixes and improvements.

* Tue Sep 27 2016 Krzysztof Kulakowski <krzysztof.kulakowski@intel.com> v1.3.0
- Introduced logging mechanism (for details please take a look at
  man/memkind.3).
- Deprecated symbols: memkind_finalize(), memkind_get_num_kind(),
  memkind_get_kind_by_partition(), memkind_get_kind_by_name(),
  memkind_partition_mmap(), memkind_get_size(), MEMKIND_ERROR_MEMALIGN,
  MEMKIND_ERROR_MALLCTL,  MEMKIND_ERROR_GETCPU, MEMKIND_ERROR_PMTT,
  MEMKIND_ERROR_TIEDISTANCE, MEMKIND_ERROR_ALIGNMENT, MEMKIND_ERROR_MALLOCX,
  MEMKIND_ERROR_REPNAME, MEMKIND_ERROR_PTHREAD, MEMKIND_ERROR_BADPOLICY,
  MEMKIND_ERROR_REPPOLICY.
- Added integration with hwloc (turned on with --with-hwloc).
- Cleanup of symbols exposed by libmemkind.so (e.g. no longer exposing libnuma
  and jemalloc symbols).
- AutoHBW files have been moved to "memkind/autohbw" directory, code has
  been refactored and tests have been added to appropriate scenarios.
- Library is now built with flags improving security (can be turned off
  with --disable-secure configure time option).
- Changed configuration of jemalloc to turn off unused features.

* Thu Aug 18 2016 Krzysztof Kulakowski <krzysztof.kulakowski@intel.com> v1.2.0
- memkind_create() and memkind_ops have been deprecated (moved to
  memkind_deprecated.h).
- Deprecated the headers from memkind/internal and added compile-time
  warnings to them (plan is to remove those from rpm packages).
- autohbw files have been moved to memkind/autohbw directory.
- API decorators (memkind_malloc_pre, memkind_malloc_post etc.) now
  need to be enabled with configure-time option --enable-decorators.
- Allocation time optimizations with up to 20% improvement.

* Wed Jul 13 2016 Krzysztof Kulakowski <krzysztof.kulakowski@intel.com> v1.1.1
- Performance improvement for memkind_free() in scenario where NULL
  was passed as kind (reduced by 60%).
- Introduced integration with Travis CI.
- Fixed issue where memory returned from calloc was not filled with zeroes
  when using memkind_pmem kinds.
- Fixed issue where interleave kinds was failing on systems without
  Transparent Huge Pages module configured.
- Resolved several issues that was causing compilation errors on some systems.
- test/test.sh is now able to detect system configuration, and run only those
  tests which requirements are meet.

src/memkind_default.c
/*
 * Copyright (C) 2014 - 2019 Intel Corporation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 1. Redistributions of source code must retain the above copyright notice(s),
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice(s),
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <memkind/internal/memkind_default.h>
#include <memkind/internal/memkind_private.h>
#include <memkind/internal/memkind_log.h>
#include <memkind/internal/tbb_wrapper.h>
#include <memkind/internal/heap_manager.h>

#include <numa.h>
#include <numaif.h>
#include <sys/mman.h>
#include <errno.h>
#include <jemalloc/jemalloc.h>
#include <stdint.h>

#ifndef MADV_NOHUGEPAGE
#define MADV_NOHUGEPAGE 15
#endif

MEMKIND_EXPORT struct memkind_ops MEMKIND_DEFAULT_OPS = {
    .create = memkind_default_create,
    .destroy = memkind_default_destroy,
    .malloc = memkind_default_malloc,
    .calloc = memkind_default_calloc,
    .posix_memalign = memkind_default_posix_memalign,
    .realloc = memkind_default_realloc,
    .free = memkind_default_free,
    .init_once = memkind_default_init_once,
    .malloc_usable_size = memkind_default_malloc_usable_size,
    .finalize = memkind_default_destroy
};

MEMKIND_EXPORT int memkind_default_create(struct memkind *kind,
                                          struct memkind_ops *ops, const char *name)
{
    int err = 0;

    kind->ops = ops;
    if (strlen(name) >= MEMKIND_NAME_LENGTH_PRIV) {
        kind->name[0] = '\0';
        err = MEMKIND_ERROR_INVALID;
    } else {
        strcpy(kind->name, name);
    }
    return err;
}

MEMKIND_EXPORT int memkind_default_destroy(struct memkind *kind)
{
    return 0;
}

MEMKIND_EXPORT void *memkind_default_malloc(struct memkind *kind, size_t size)
{
    if(MEMKIND_UNLIKELY(size_out_of_bounds(size))) {
        return NULL;
    }
    return jemk_malloc(size);
}

MEMKIND_EXPORT void *memkind_default_calloc(struct memkind *kind, size_t num,
                                            size_t size)
{
    if(MEMKIND_UNLIKELY(size_out_of_bounds(num) || size_out_of_bounds(size))) {
        return NULL;
    }
    return jemk_calloc(num, size);
}

MEMKIND_EXPORT int memkind_default_posix_memalign(struct memkind *kind,
                                                  void **memptr, size_t alignment, size_t size)
{
    if(MEMKIND_UNLIKELY(size_out_of_bounds(size))) {
        *memptr = NULL;
        return 0;
    }
    return jemk_posix_memalign(memptr, alignment, size);
}

MEMKIND_EXPORT void *memkind_default_realloc(struct memkind *kind, void *ptr,
                                             size_t size)
{
    if(MEMKIND_UNLIKELY(size_out_of_bounds(size))) {
        jemk_free(ptr);
        return NULL;
    }
    return jemk_realloc(ptr, size);
}

MEMKIND_EXPORT void memkind_default_free(struct memkind *kind, void *ptr)
{
    jemk_free(ptr);
}

MEMKIND_EXPORT size_t memkind_default_malloc_usable_size(struct memkind *kind,
                                                         void *ptr)
{
    return jemk_malloc_usable_size(ptr);
}

MEMKIND_EXPORT int memkind_default_get_defrag_hint(void* ptr, int *bin_util, int *run_util)
{
  return je_get_defrag_hint(ptr,bin_util,run_util);
}

MEMKIND_EXPORT void *memkind_default_mmap(struct memkind *kind, void *addr,
                                          size_t size)
{
    void *result = MAP_FAILED;
    int err = 0;
    int flags;

    if (kind->ops->get_mmap_flags) {
        err = kind->ops->get_mmap_flags(kind, &flags);
    } else {
        err = memkind_default_get_mmap_flags(kind, &flags);
    }
    if (MEMKIND_LIKELY(!err)) {
        result = mmap(addr, size, PROT_READ | PROT_WRITE, flags, -1, 0);
        if (result == MAP_FAILED) {
            log_err("syscall mmap() returned: %p", result);
            return result;
        }
    }
    if (kind->ops->mbind) {
        err = kind->ops->mbind(kind, result, size);
        if (err) {
            munmap(result, size);
            result = MAP_FAILED;
        }
    }
    if (kind->ops->madvise) {
        err = kind->ops->madvise(kind, result, size);
        if (err) {
            munmap(result, size);
            result = MAP_FAILED;
        }
    }
    return result;
}

MEMKIND_EXPORT int memkind_nohugepage_madvise(struct memkind *kind, void *addr,
                                              size_t size)
{
    int err = madvise(addr, size, MADV_NOHUGEPAGE);

    //checking if EINVAL was returned due to lack of THP support in kernel
    if ((err == EINVAL) && (((uintptr_t) addr & (uintptr_t) 0xfff) == 0) &&
        (size > 0)) {
        return 0;
    }
    if (MEMKIND_UNLIKELY(err)) {
        log_err("syscall madvise() returned: %d", err);
    }
    return err;
}

MEMKIND_EXPORT int memkind_default_mbind(struct memkind *kind, void *ptr,
                                         size_t size)
{
    nodemask_t nodemask;
    int err = 0;
    int mode;

    if (MEMKIND_UNLIKELY(kind->ops->get_mbind_nodemask == NULL ||
                         kind->ops->get_mbind_mode == NULL)) {
        log_err("memkind_ops->mbind_mode or memkind_ops->bind_nodemask is NULL.");
        return MEMKIND_ERROR_BADOPS;
    }
    err = kind->ops->get_mbind_nodemask(kind, nodemask.n, NUMA_NUM_NODES);
    if (MEMKIND_UNLIKELY(err)) {
        return err;
    }
    err = kind->ops->get_mbind_mode(kind, &mode);
    if (MEMKIND_UNLIKELY(err)) {
        return err;
    }
    err = mbind(ptr, size, mode, nodemask.n, NUMA_NUM_NODES, 0);
    if (MEMKIND_UNLIKELY(err)) {
        log_err("syscall mbind() returned: %d", err);
        return MEMKIND_ERROR_MBIND;
    }
    return err;
}

MEMKIND_EXPORT int memkind_default_get_mmap_flags(struct memkind *kind,
                                                  int *flags)
{
    *flags = MAP_PRIVATE | MAP_ANONYMOUS;
    return 0;
}

MEMKIND_EXPORT int memkind_default_get_mbind_nodemask(struct memkind *kind,
                                                      unsigned long *nodemask,
                                                      unsigned long maxnode)
{
    struct bitmask nodemask_bm = {maxnode, nodemask};
    copy_bitmask_to_bitmask(numa_all_nodes_ptr, &nodemask_bm);
    return 0;
}

MEMKIND_EXPORT int memkind_default_get_mbind_mode(struct memkind *kind,
                                                  int *mode)
{
    *mode = MPOL_BIND;
    return 0;
}

MEMKIND_EXPORT int memkind_preferred_get_mbind_mode(struct memkind *kind,
                                                    int *mode)
{
    *mode = MPOL_PREFERRED;
    return 0;
}

MEMKIND_EXPORT int memkind_interleave_get_mbind_mode(struct memkind *kind,
                                                     int *mode)
{
    *mode = MPOL_INTERLEAVE;
    return 0;
}

MEMKIND_EXPORT int memkind_posix_check_alignment(struct memkind *kind,
                                                 size_t alignment)
{
    int err = 0;
    if ((alignment < sizeof(void *)) ||
        (((alignment - 1) & alignment) != 0)) {
        err = EINVAL;
    }
    return err;
}

MEMKIND_EXPORT void memkind_default_init_once(void)
{
    heap_manager_init(MEMKIND_DEFAULT);
}
